server/src/http/broadcaster.rs

```diff
@@ -0,0 +1,207 @@
+
+            group.clients = ok_clients;
+        }
+    }
+
+    /// Registers client with broadcaster, returning an SSE response body.
+    /// The current impl takes the feature set as input and sends it to the client as a connected event.
+    ///
+    /// The commented-out arguments are what we'll need to store per client so
+    /// that we can properly filter / format the feature response when they get
+    /// updates later.
+    pub async fn connect(
+        &self,
+        token: EdgeToken,
+        filter_set: Query<FeatureFilters>,
+        query: unleash_types::client_features::Query,
+    ) -> EdgeResult<Sse<InfallibleStream<ReceiverStream<sse::Event>>>> {
+        let (tx, rx) = mpsc::channel(10);
+
+        let features = &self
+            .resolve_features(&token, filter_set.clone(), query.clone())
+            .await?;
+
+        tx.send(
+            sse::Data::new_json(features)?
+                .event("unleash-connected")
+                .into(),
+        )
+        .await?;
+
+        self.active_connections
+            .entry(QueryWrapper { query })
+            .and_modify(|group| {
+                group.clients.push(tx.clone());
+            })
+            .or_insert(ClientGroup {
+                clients: vec![tx.clone()],
+                filter_set,
+                token,
+            });
+
+        Ok(Sse::from_infallible_receiver(rx))
+    }
+
+    fn get_query_filters(
+        filter_query: Query<FeatureFilters>,
+        token: &EdgeToken,
+    ) -> FeatureFilterSet {
+        let query_filters = filter_query.into_inner();
+
+        let filter_set = if let Some(name_prefix) = query_filters.name_prefix {
+            FeatureFilterSet::from(Box::new(name_prefix_filter(name_prefix)))
+        } else {
+            FeatureFilterSet::default()
+        }
+        .with_filter(project_filter(token));
+        filter_set
+    }
+
+    async fn resolve_features(
+        &self,
+        validated_token: &EdgeToken,
+        filter_set: Query<FeatureFilters>,
+        query: FlagQuery,
+    ) -> EdgeJsonResult<ClientFeatures> {
+        let filter_set = Broadcaster::get_query_filters(filter_set.clone(), validated_token);
+
+        let features = self
+            .features_cache
+            .get(&cache_key(validated_token))
+            .map(|client_features| filter_client_features(&client_features, &filter_set));
+
+        match features {
+            Some(features) => Ok(Json(ClientFeatures {
+                query: Some(query),
+                ..features
+            })),
+            // Note: this is a simplification for now, using the following assumptions:
+            // 1. We'll only allow streaming in strict mode
+            // 2. We'll check whether the token is subsumed *before* trying to add it to the broadcaster
+            // If both of these are true, then we should never hit this case (if Thomas's understanding is correct).
+            None => todo!(),
```

Comment:
I agree with the assumptions, but should we return an error instead if this is ever reached?

Reply:
yeah, let's do it. better to be on the safe side 🤷🏼